extensions/email/src/main/java/org/mule/extension/email/api/retriever/RetrieverConnection.java

```diff
@@ -0,0 +1,134 @@
+import org.mule.runtime.api.connection.ConnectionValidationResult;
+
+import java.util.Map;
+import java.util.Properties;
+
+import javax.mail.Authenticator;
+import javax.mail.Folder;
+import javax.mail.MessagingException;
+import javax.mail.Session;
+import javax.mail.Store;
+
+/**
+ * A connection with a mail server for retrieving emails from an specific folder.
+ *
+ * @since 4.0
+ */
+public class RetrieverConnection implements EmailConnection
+{
+
+    private final Session session;
+    private final Store store;
+    private final Folder folder;
+
+    /**
+     * Creates a new instance of the of the connection.
+     *
+     * @param protocol the desired protocol to use.
+     * @param username the username to establish the connection.
+     * @param password the password corresponding to the {@code username}.
+     * @param host the host name of the mail server
+     * @param port the port number of the mail server.
+     * @param properties additional custom properties.
+     * @param authenticator custom authenticator to perform the connection with the mail server.
+     * @param folder the folder to be opened in order to retrieve emails.
+     */
+    public RetrieverConnection(String protocol,
+                               String username,
+                               String password,
+                               String host,
+                               String port,
+                               Map<String, String> properties,
+                               Authenticator authenticator,
+                               String folder) throws EmailConnectionException
+    {
+        Properties sessionProperties = EmailProperties.get(protocol, host, port, properties);
+        this.session = Session.getInstance(sessionProperties, authenticator);
+        try
+        {
+            this.store = session.getStore(protocol);
+            this.store.connect(username, password);
+            this.folder = store.getFolder(folder);
+        }
+        catch (MessagingException e)
+        {
+            throw new EmailConnectionException("Error while acquiring connection with the " + protocol + "store", e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void disconnect()
+    {
+        try
+        {
+            store.close();
+            if (folder.isOpen())
+            {
+                folder.close(false);
+            }
+        }
+        catch (MessagingException e)
+        {
+            throw new EmailException("Error while disconnecting", e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Session getSession()
+    {
+        return session;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ConnectionValidationResult validate()
+    {
+        String errorMessage = "";
+        return session != null && store.isConnected() ? success() : failure(errorMessage, UNKNOWN, new EmailConnectionException(errorMessage));
```

Comment:
we probably need to add a fault code `DISCONNECTED` since `store.isConnected() == false` and `store == null` are two different things